Add --json option to version command

CommanderRoot opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Add --json option to version command

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

#4547 raised the issue of not having the option to output version information as json

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

CommanderRoot committed:
command version: add json output option

CommanderRoot committed:
add changelog/unreleases for issue-4547

MichaelEischer commented on cmd/restic/cmd_version.go:
```diff
@@ -21,8 +22,31 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
+				GoTarget:  runtime.GOOS + "/" + runtime.GOARCH,
+			}
+
+			jsonB, err := json.Marshal(jsonS)
```
I'd prefer to use `return json.NewEncoder(globalOptions.stdout).Encode(jsonS)`  (see `cmd_init.go`).

MichaelEischer commented on changelog/unreleased/issue-4547:
```diff
@@ -0,0 +1,6 @@
+Restic now supports outputting restic version and used go version and platform
+target via json when using the version command
+
+https://github.com/restic/restic/issues/4547
```
Please also add a link to this PR.

MichaelEischer commented on changelog/unreleased/issue-4547:
```diff
@@ -0,0 +1,6 @@
+Enhancement: add --json option to version command.
```
Please change this to 
```
Enhancement: Add support for `--json` option to `version` command
```

MichaelEischer commented on changelog/unreleased/issue-4547:
```diff
@@ -0,0 +1,6 @@
+Enhancement: add --json option to version command.
+
+Restic now supports outputting restic version and used go version and platform
+target via json when using the version command
```
There's a point missing.

MichaelEischer reviewed (changes requested):
I've added a few comments below. Can you add the `version` command to the JSON format documentation at https://github.com/restic/restic/blob/master/doc/075_scripting.rst#json-output ?

CommanderRoot committed:
Update wording on changelog entry

CommanderRoot committed:
Add version command output to JSON format documentation

CommanderRoot committed:
Use different function to be more consistent with other code

CommanderRoot commented:
I have made all requested changes and added the json foramt documentation

rawtaz commented on cmd/restic/cmd_version.go:
```diff
@@ -21,8 +22,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
+			type jsonVersion struct {
+				Version   string `json:"version"`
+				GoVersion string `json:"go_version"`
+				GoTarget  string `json:"go_target"`
```
We discussed this a bit and would like to replace `go_target` with two separate keys instead, `go_os` and `go_arch` (which corresponds to `runtime.GOOS` and `runtime.GOARCH`). This way, users can more easily get to the individual pieces of data instead of having to split it up separately.

rawtaz commented on changelog/unreleased/issue-4547:
```diff
@@ -0,0 +1,7 @@
+Enhancement: Add support for `--json` option to `version` command
+
+Restic now supports outputting restic version and used go version and platform
+target via json when using the version command.
```
s/json/JSON/ :)

rawtaz replied:
Perhaps also change "go version and platform target" to "go version, OS and architecture" to match the other requested change.

rawtaz commented on cmd/restic/cmd_version.go:
```diff
@@ -21,8 +22,29 @@ Exit status is 0 if the command was successful, and non-zero if there was any er
+
+			err := json.NewEncoder(globalOptions.stdout).Encode(jsonS)
+			if err != nil {
+				Warnf("Encode failed: %v\n", err)
```
The `diff` and `ls` commands output this error as `JSON encode failed: %v\n`. Can we please change it to that so it's more coherent (and also clear that it's about JSON)?

rawtaz commented on doc/075_scripting.rst:
```diff
@@ -576,3 +576,17 @@ The snapshots command returns a single JSON object.
++----------------+--------------------+
+| ``go_version`` | Go compile version |
++----------------+--------------------+
+| ``go_target``  | Go target platform |
```
This needs to be updated as well, to match the other requested change. Thanks!

rawtaz reviewed (changes requested):
Just some small-ish comments :)

CommanderRoot committed:
Split `go_target` into `go_os` and `go_arch`

CommanderRoot commented:
Based on the feedback I have split the `go_target` into separate values (`go_os` and `go_arch`). I have also updated the documentation to reflect this change

rawtaz reviewed (approved):
LGTM.

MichaelEischer reviewed (approved):
LGTM. Thanks!
